File: foolish-core-java/src/main/java/org/foolish/fvm/ubc/RegexpSearchFiroe.java
package org.foolish.fvm.ubc;

import org.foolish.ast.AST;
import org.foolish.ast.SearchOperator;
import java.util.Optional;

public class RegexpSearchFiroe extends AbstractSearchFiroe implements Constanicable {
    private final RegexpSearcher searcher;

    public RegexpSearchFiroe(AST.RegexpSearchExpr searchExpr) {
        super(searchExpr, searchExpr.operator());
        this.searcher = new RegexpSearcher(searchExpr.pattern());
    }

    public RegexpSearchFiroe(Either<AST, String> source, String pattern, SearchOperator operator) {
        super(source, operator);
        this.searcher = new RegexpSearcher(pattern);
    }
    
    /**
     * Copy constructor for cloneConstanic.
     */
    protected RegexpSearchFiroe(RegexpSearchFiroe original, FIR newParent) {
        super(original, newParent);
        this.searcher = original.searcher;
    }

    @Override
    protected void initialize() {
        super.initialize();
        if (ast instanceof AST.RegexpSearchExpr searchExpr) {
            storeExprs(searchExpr.anchor());
        }
    }
    
    @Override 
    protected SearchCursor createCursor(BraneFiroe target) {
        ReadOnlyBraneMemory memory = target.getBraneMemory();
        int size = memory.size();

        // Check if GLOBAL search (searches up) or LOCAL (bounded)
        boolean isGlobal = (operator == SearchOperator.REGEXP_GLOBAL || operator == SearchOperator.REGEXP_FORWARD_GLOBAL);
        boolean braneBound = !isGlobal;

        // Direction logic: FORWARD means start at 0, otherwise standard usually reverse
        boolean forward = (operator == SearchOperator.REGEXP_FORWARD_LOCAL || operator == SearchOperator.REGEXP_FORWARD_GLOBAL);
        
        int startIndex = forward ? 0 : Math.max(0, size - 1);
        
        return new SearchCursor(new FoolishCursor(target, startIndex), forward, true, braneBound);
    }

    @Override
    protected FIR executeSearch(SearchCursor cursor) {
         return searcher.search(cursor).orElse(new MissingFiroe());
    }
    
    @Override
    public String toString() {
        if (ast != null) return ast.toString();
        return searcher.getRawPattern();
    }
    
    public String getPattern() {
       return searcher.getPattern();
    }

    @Override
    protected FIR cloneConstanic(FIR newParent, Optional<Nyes> targetNyes) {
        if (!isConstanic()) {
            throw new IllegalStateException(
                formatErrorMessage("cloneConstanic can only be called on CONSTANIC or CONSTANT FIRs, " +
                    "but this FIR is in state: " + getNyes()));
        }

        if (isConstant()) {
            return this;
        }
        
        RegexpSearchFiroe copy = new RegexpSearchFiroe(this, newParent);
        if (targetNyes.isPresent()) {
            copy.nyes = targetNyes.get();
        } else {
            copy.nyes = this.nyes;
        }
        return copy;
    }
}

File: foolish-core-scala/src/main/scala/org/foolish/fvm/scubc/RegexpSearchFiroe.scala
package org.foolish.fvm.scubc

import org.foolish.ast.{AST, SearchOperator}

import java.util.Optional

/**
 * RegexpSearchFiroe performs a regular expression search on a brane.
 */
class RegexpSearchFiroe(regexpSearch: AST.RegexpSearchExpr) extends AbstractSearchFiroe(regexpSearch, regexpSearch.operator()) {
  private var pattern: String = regexpSearch.pattern()

  override protected def initialize(): Unit = {
    super.initialize()
    val searchExpr = ast.asInstanceOf[AST.RegexpSearchExpr]
    enqueueExprs(searchExpr.anchor())
  }

  override protected def executeSearch(target: BraneFiroe): FIR = {
    val query = new BraneMemory.RegexpQuery(pattern)
    val targetMemory = target.braneMemory

    val result = operator match {
      case SearchOperator.REGEXP_LOCAL =>
        // Backward search: search from end to start (find last match)
        val searchFrom = targetMemory.size - 1
        targetMemory.getLocal(query, searchFrom)
      case SearchOperator.REGEXP_FORWARD_LOCAL =>
        // Forward search: search from start to end (find first match)
        val searchFrom = 0
        targetMemory.getLocalForward(query, searchFrom)
      case SearchOperator.REGEXP_GLOBAL =>
        // Global backward search (find-all, not yet fully implemented)
        val searchFrom = targetMemory.size - 1
        targetMemory.get(query, searchFrom)
      case _ => throw new IllegalStateException("Unknown regexp operator: " + operator)
    }

    if (result.isDefined) result.get._2 else new NKFiroe()
  }

  override def toString: String = ast.toString

  /**
   * Copy constructor for cloneConstanic.
   * Resets search state so the search can be re-executed in a new context.
   */
  private def this(original: RegexpSearchFiroe, newParent: FIR) =
    this(original.ast.asInstanceOf[AST.RegexpSearchExpr])
    setParentFir(newParent)
    pattern = original.pattern
    // Copy braneMemory contents with cloned items
    original.braneMemory.stream.foreach { fir =>
      val cloned = fir.asInstanceOf[Constanicable].cloneConstanic(this, Some(Nyes.INITIALIZED))
      braneMemory.put(cloned)
      indexLookup.put(cloned, braneMemory.size - 1)
      // For nested FiroeWithBraneMind instances, set up parent memory link
      if cloned.isInstanceOf[FiroeWithBraneMind] then
        cloned.asInstanceOf[FiroeWithBraneMind].ordinated = false
        cloned.asInstanceOf[FiroeWithBraneMind].ordinateToParentBraneMind(this, indexLookup.size - 1)
    }

  override def cloneConstanic(newParent: FIR, targetNyes: Option[Nyes]): FIR =
    if !isConstanic then
      throw IllegalStateException(
        s"cloneConstanic can only be called on CONSTANIC or CONSTANT FIRs, " +
        s"but this FIR is in state: ${getNyes}")
    if isConstant then
      return this  // Share CONSTANT searches
    val copy = new RegexpSearchFiroe(this, newParent)
    targetNyes.foreach(copy.setNyes)
    copy
}
